File: android/src/test/java/com/digia/engage/webengage/WebEngagePluginTest.kt
package com.digia.engage.webengage

import com.digia.engage.DigiaCEPDelegate
import com.digia.engage.DigiaExperienceEvent
import com.digia.engage.InAppPayload
import org.junit.Assert.assertEquals
import org.junit.Test
import java.lang.ref.WeakReference

class WebEngagePluginTest {

    @Test
    fun `dispatches mapped inapp payload to delegate`() {
        val bridge = FakeBridge()
        val config = WebEngagePluginConfig(diagnosticsEnabled = false)
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(config),
            config = config,
        )
        val delegate = FakeDelegate()

        plugin.setup(delegate)
        bridge.emitInApp(
            mapOf(
                "experimentId" to "exp-1",
                "command" to "SHOW_DIALOG",
                "viewId" to "welcome_modal",
                "screenId" to "home",
                "args" to mapOf("title" to "Hello"),
            ),
        )

        assertEquals(1, delegate.triggered.size)
        assertEquals("exp-1", delegate.triggered.first().id)
    }

    @Test
    fun `dispatches invalidation to delegate`() {
        val bridge = FakeBridge()
        val config = WebEngagePluginConfig(diagnosticsEnabled = false)
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(config),
            config = config,
        )
        val delegate = FakeDelegate()

        plugin.setup(delegate)
        bridge.emitInvalidate("exp-7")

        assertEquals(listOf("exp-7"), delegate.invalidated)
    }

    @Test
    fun `dispatches inline payload to delegate`() {
        val bridge = FakeBridge()
        val config = WebEngagePluginConfig(diagnosticsEnabled = false)
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(config),
            config = config,
        )
        val delegate = FakeDelegate()

        plugin.setup(delegate)
        bridge.emitInline(
            campaignId = "cmp-1",
            targetViewId = "hero_slot",
            customData = mapOf("componentId" to "hero_component"),
            metadata = mapOf("variationId" to "var-1", "propertyId" to "hero_slot"),
        )

        assertEquals(1, delegate.triggered.size)
        assertEquals("cmp-1:hero_slot", delegate.triggered.first().id)
    }

    @Test
    fun `drops payloads when delegate reference is unavailable`() {
        val bridge = FakeBridge()
        val config = WebEngagePluginConfig(diagnosticsEnabled = false)
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(config),
            config = config,
        )
        val delegate = FakeDelegate()

        plugin.setup(delegate)
        clearDelegateReference(plugin)
        bridge.emitInApp(
            mapOf(
                "experimentId" to "exp-1",
                "command" to "SHOW_DIALOG",
                "viewId" to "welcome_modal",
            ),
        )
        bridge.emitInline(
            campaignId = "cmp-1",
            targetViewId = "hero_slot",
            customData = mapOf("componentId" to "hero_component"),
            metadata = emptyMap(),
        )
        bridge.emitInvalidate("exp-1")

        assertEquals(emptyList<InAppPayload>(), delegate.triggered)
        assertEquals(emptyList<String>(), delegate.invalidated)
    }

    @Test
    fun `dispatches fallback dialog for non digia payload in suppress all mode`() {
        val bridge = FakeBridge()
        val config = WebEngagePluginConfig(
            suppressionMode = SuppressionMode.SUPPRESS_ALL,
            diagnosticsEnabled = false,
            forcedDialogComponentId = "coupon_nudge-b6dByb",
        )
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(config),
            config = config,
        )
        val delegate = FakeDelegate()

        plugin.setup(delegate)
        bridge.emitInApp(
            mapOf(
                "experimentId" to "exp-2",
                "title" to "plain webengage campaign",
            ),
        )

        assertEquals(1, delegate.triggered.size)
        assertEquals("SHOW_DIALOG", delegate.triggered.first().content["command"])
        assertEquals("coupon_nudge-b6dByb", delegate.triggered.first().content["viewId"])
    }

    @Test
    fun `notifyEvent maps only nudge impression to webengage notification system events`() {
        val bridge = FakeBridge()
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(WebEngagePluginConfig(diagnosticsEnabled = false)),
            config = WebEngagePluginConfig(diagnosticsEnabled = false),
        )
        plugin.setup(FakeDelegate())

        val payload = InAppPayload(
            id = "exp-9",
            content = mapOf("command" to "SHOW_DIALOG"),
            cepContext = mapOf(
                "experimentId" to "exp-9",
                "variationId" to "var-9",
            ),
        )

        plugin.notifyEvent(DigiaExperienceEvent.Impressed, payload)
        plugin.notifyEvent(DigiaExperienceEvent.Clicked("cta_apply"), payload)
        plugin.notifyEvent(DigiaExperienceEvent.Dismissed, payload)

        assertEquals(1, bridge.trackedSystemEvents.size)
        assertEquals("notification_view", bridge.trackedSystemEvents[0].eventName)
    }

    @Test
    fun `notifyEvent maps inline events to personalization system events`() {
        val bridge = FakeBridge()
        val plugin = WebEngagePlugin.createForTest(
            bridge = bridge,
            mapper = WebEngagePayloadMapper(WebEngagePluginConfig(diagnosticsEnabled = false)),
            config = WebEngagePluginConfig(diagnosticsEnabled = false),
        )
        plugin.setup(FakeDelegate())

        val payload = InAppPayload(
            id = "cmp-1:hero_slot",
            content = mapOf(
                "command" to "SHOW_INLINE",
                "placementKey" to "hero_slot",
                "args" to mapOf("foo" to "bar"),
            ),
            cepContext = mapOf(
                "campaignId" to "cmp-1",
                "variationId" to "var-1",
                "propertyId" to "hero_slot",
            ),
        )

        plugin.notifyEvent(DigiaExperienceEvent.Impressed, payload)
        plugin.notifyEvent(DigiaExperienceEvent.Clicked("cta_inline"), payload)
        plugin.notifyEvent(DigiaExperienceEvent.Dismissed, payload)

        assertEquals(2, bridge.trackedSystemEvents.size)
        assertEquals("app_personalization_view", bridge.trackedSystemEvents[0].eventName)
        assertEquals("app_personalization_click", bridge.trackedSystemEvents[1].eventName)
        assertEquals("hero_slot", bridge.trackedSystemEvents[0].systemData["p_id"])
        assertEquals("cta_inline", bridge.trackedSystemEvents[1].systemData["actionId"])
        assertEquals("bar", bridge.trackedSystemEvents[1].eventData["foo"])
    }

    private fun clearDelegateReference(plugin: WebEngagePlugin) {
        val field = WebEngagePlugin::class.java.getDeclaredField("delegateRef")
        field.isAccessible = true
        @Suppress("UNCHECKED_CAST")
        val weakRef = field.get(plugin) as? WeakReference<DigiaCEPDelegate>
        weakRef?.clear()
    }

    private class FakeDelegate : DigiaCEPDelegate {
        val triggered = mutableListOf<InAppPayload>()
        val invalidated = mutableListOf<String>()

        override fun onCampaignTriggered(payload: InAppPayload) {
            triggered += payload
        }

        override fun onCampaignInvalidated(campaignId: String) {
            invalidated += campaignId
        }
    }

    private class FakeBridge : WebEngageBridge {
        private var inAppCallback: ((Map<String, Any?>) -> Unit)? = null
        private var invalidateCallback: ((String) -> Unit)? = null
        private var inlineCallback: ((String, String, Map<String, Any?>, Map<String, Any?>) -> Unit)? = null
        val trackedSystemEvents = mutableListOf<TrackedSystemEvent>()

        override fun registerInAppListener(
            onPayload: (Map<String, Any?>) -> Unit,
            onInvalidate: (String) -> Unit,
        ) {
            inAppCallback = onPayload
            invalidateCallback = onInvalidate
        }

        override fun registerInlineListener(
            onInlineCampaign: (
                campaignId: String,
                targetViewId: String,
                customData: Map<String, Any?>,
                metadata: Map<String, Any?>,
            ) -> Unit,
        ) {
            inlineCallback = onInlineCampaign
        }

        override fun trackSystemEvent(
            eventName: String,
            systemData: Map<String, Any?>,
            eventData: Map<String, Any?>,
        ) {
            trackedSystemEvents += TrackedSystemEvent(
                eventName = eventName,
                systemData = systemData,
                eventData = eventData,
            )
        }

        override fun unregisterInAppListener() {
            inAppCallback = null
            invalidateCallback = null
        }

        override fun unregisterInlineListener() {
            inlineCallback = null
        }

        override fun navigateScreen(name: String) = Unit

        override fun isAvailable(): Boolean = true

        fun emitInApp(payload: Map<String, Any?>) {
            inAppCallback?.invoke(payload)
        }

        fun emitInvalidate(campaignId: String) {
            invalidateCallback?.invoke(campaignId)
        }

        fun emitInline(
            campaignId: String,
            targetViewId: String,
            customData: Map<String, Any?>,
            metadata: Map<String, Any?>,
        ) {
            inlineCallback?.invoke(campaignId, targetViewId, customData, metadata)
        }
    }

    private data class TrackedSystemEvent(
        val eventName: String,
        val systemData: Map<String, Any?>,
        val eventData: Map<String, Any?>,
    )
}

File: react-native/__tests__/WebEngagePlugin.test.ts
import { WebEngagePlugin } from '../src/WebEngagePlugin';
import { DigiaExperienceEvent } from '../src/types';
import type { DigiaCEPDelegate, InAppPayload } from '../src/types';
import type { WebEngageBridge, WEInAppData } from '../src/WebEngageBridge';
import { SuppressionMode } from '../src/config';

// ─── FakeBridge ───────────────────────────────────────────────────────────────

interface TrackedSystemEvent {
    eventName: string;
    systemData: Record<string, unknown>;
    eventData: Record<string, unknown>;
}

class FakeBridge implements WebEngageBridge {
    private _preparedCb: ((data: WEInAppData) => void) | null = null;
    private _dismissedCb: ((id: string) => void) | null = null;
    readonly trackedSystemEvents: TrackedSystemEvent[] = [];
    readonly navigatedScreens: string[] = [];
    private _available = true;

    registerCallbacks(callbacks: {
        onInAppPrepared: (data: WEInAppData) => void;
        onInAppDismissed: (campaignId: string) => void;
    }): void {
        this._preparedCb = callbacks.onInAppPrepared;
        this._dismissedCb = callbacks.onInAppDismissed;
    }

    unregisterCallbacks(): void {
        this._preparedCb = null;
        this._dismissedCb = null;
    }

    navigateScreen(name: string): void {
        this.navigatedScreens.push(name);
    }

    trackSystemEvent(
        eventName: string,
        systemData: Record<string, unknown>,
        eventData: Record<string, unknown>,
    ): void {
        this.trackedSystemEvents.push({ eventName, systemData, eventData });
    }

    isAvailable(): boolean {
        return this._available;
    }

    // ─── Test helpers ───────────────────────────────────────────────────────────

    emitInApp(data: WEInAppData): void {
        this._preparedCb?.(data);
    }

    emitDismissed(campaignId: string): void {
        this._dismissedCb?.(campaignId);
    }

    setUnavailable(): void {
        this._available = false;
    }
}

// ─── FakeDelegate ──────────────────────────────────────────────────────────────

class FakeDelegate implements DigiaCEPDelegate {
    readonly triggered: InAppPayload[] = [];
    readonly invalidated: string[] = [];

    onCampaignTriggered(payload: InAppPayload): void {
        this.triggered.push(payload);
    }

    onCampaignInvalidated(campaignId: string): void {
        this.invalidated.push(campaignId);
    }
}

// ─── Factory ──────────────────────────────────────────────────────────────────

function makePlugin(bridge: FakeBridge, configOverrides = {}) {
    return new WebEngagePlugin({ bridge, config: configOverrides });
}

// ─── Tests ────────────────────────────────────────────────────────────────────

describe('WebEngagePlugin — campaign dispatch', () => {
    it('dispatches mapped in-app payload to delegate', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        bridge.emitInApp({
            experimentId: 'exp-1',
            command: 'SHOW_DIALOG',
            viewId: 'welcome_modal',
            screenId: 'home',
            args: { title: 'Hello' },
        });

        expect(delegate.triggered).toHaveLength(1);
        expect(delegate.triggered[0].id).toBe('exp-1');
        expect(delegate.triggered[0].content['command']).toBe('SHOW_DIALOG');
        expect(delegate.triggered[0].content['viewId']).toBe('welcome_modal');
    });

    it('dispatches invalidation to delegate', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        bridge.emitDismissed('exp-7');

        expect(delegate.invalidated).toEqual(['exp-7']);
    });

    it('drops payloads when campaign id is missing', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        bridge.emitInApp({ command: 'SHOW_DIALOG', viewId: 'modal' }); // no experimentId

        expect(delegate.triggered).toHaveLength(0);
    });

    it('ignores empty dismiss campaignId', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        bridge.emitDismissed('');

        expect(delegate.invalidated).toHaveLength(0);
    });

    it('stops dispatching after teardown', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        plugin.teardown();
        bridge.emitInApp({
            experimentId: 'exp-1',
            command: 'SHOW_DIALOG',
            viewId: 'welcome_modal',
        });
        bridge.emitDismissed('exp-1');

        expect(delegate.triggered).toHaveLength(0);
        expect(delegate.invalidated).toHaveLength(0);
    });

    it('dispatches fallback dialog for non-Digia payload in SUPPRESS_ALL mode', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge, {
            suppressionMode: SuppressionMode.SUPPRESS_ALL,
            forcedDialogComponentId: 'coupon_nudge-b6dByb',
        });
        const delegate = new FakeDelegate();

        plugin.setup(delegate);
        bridge.emitInApp({ experimentId: 'exp-2', title: 'plain webengage campaign' });

        expect(delegate.triggered).toHaveLength(1);
        expect(delegate.triggered[0].id).toBe('exp-2:forced_dialog');
        expect(delegate.triggered[0].content['command']).toBe('SHOW_DIALOG');
        expect(delegate.triggered[0].content['viewId']).toBe('coupon_nudge-b6dByb');
    });
});

// ─── In-app system events ─────────────────────────────────────────────────────

describe('WebEngagePlugin — notification_* system events', () => {
    const nudgePayload: InAppPayload = {
        id: 'exp-9',
        content: { command: 'SHOW_DIALOG', viewId: 'v1' },
        cepContext: { experimentId: 'exp-9', variationId: 'var-9' },
    };

    it('dispatches notification_view on Impressed', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.impressed(), nudgePayload);

        expect(bridge.trackedSystemEvents).toHaveLength(1);
        expect(bridge.trackedSystemEvents[0].eventName).toBe('notification_view');
        expect(bridge.trackedSystemEvents[0].systemData['experiment_id']).toBe('exp-9');
        expect(bridge.trackedSystemEvents[0].systemData['id']).toBe('var-9');
    });

    it('dispatches notification_click with call_to_action on Clicked', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.clicked('cta_apply'), nudgePayload);

        expect(bridge.trackedSystemEvents[0].eventName).toBe('notification_click');
        expect(bridge.trackedSystemEvents[0].systemData['call_to_action']).toBe('cta_apply');
    });

    it('omits call_to_action when elementId is absent on Clicked', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.clicked(), nudgePayload);

        expect(bridge.trackedSystemEvents[0].eventName).toBe('notification_click');
        expect(bridge.trackedSystemEvents[0].systemData['call_to_action']).toBeUndefined();
    });

    it('dispatches notification_close on Dismissed', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.dismissed(), nudgePayload);

        expect(bridge.trackedSystemEvents[0].eventName).toBe('notification_close');
    });

    it('dispatches all three in-app events in order', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.impressed(), nudgePayload);
        plugin.notifyEvent(DigiaExperienceEvent.clicked('cta_apply'), nudgePayload);
        plugin.notifyEvent(DigiaExperienceEvent.dismissed(), nudgePayload);

        const names = bridge.trackedSystemEvents.map((e) => e.eventName);
        expect(names).toEqual(['notification_view', 'notification_click', 'notification_close']);
    });
});

// ─── Inline system events ─────────────────────────────────────────────────────

describe('WebEngagePlugin — app_personalization_* system events', () => {
    const inlinePayload: InAppPayload = {
        id: 'cmp-1:hero_slot',
        content: {
            type: 'inline',
            placementKey: 'hero_slot',
            args: { foo: 'bar' },
        },
        cepContext: {
            campaignId: 'cmp-1',
            variationId: 'var-1',
            propertyId: 'hero_slot',
        },
    };

    it('dispatches app_personalization_view on Impressed', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.impressed(), inlinePayload);

        expect(bridge.trackedSystemEvents[0].eventName).toBe('app_personalization_view');
        expect(bridge.trackedSystemEvents[0].systemData['p_id']).toBe('hero_slot');
        expect(bridge.trackedSystemEvents[0].systemData['experiment_id']).toBe('cmp-1');
        expect(bridge.trackedSystemEvents[0].systemData['id']).toBe('var-1');
    });

    it('dispatches app_personalization_click with cta and args on Clicked', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.clicked('cta_inline'), inlinePayload);

        expect(bridge.trackedSystemEvents[0].eventName).toBe('app_personalization_click');
        expect(bridge.trackedSystemEvents[0].systemData['call_to_action']).toBe('cta_inline');
        expect(bridge.trackedSystemEvents[0].eventData['foo']).toBe('bar');
    });

    it('does not dispatch any event for inline Dismissed (no WE system event exists)', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.dismissed(), inlinePayload);

        expect(bridge.trackedSystemEvents).toHaveLength(0);
    });

    it('dispatches exactly 2 events for impressed + clicked (skips dismissed)', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.notifyEvent(DigiaExperienceEvent.impressed(), inlinePayload);
        plugin.notifyEvent(DigiaExperienceEvent.clicked('cta_inline'), inlinePayload);
        plugin.notifyEvent(DigiaExperienceEvent.dismissed(), inlinePayload);

        expect(bridge.trackedSystemEvents).toHaveLength(2);
        expect(bridge.trackedSystemEvents.map((e) => e.eventName)).toEqual([
            'app_personalization_view',
            'app_personalization_click',
        ]);
    });
});

// ─── Screen forwarding ────────────────────────────────────────────────────────

describe('WebEngagePlugin — screen forwarding', () => {
    it('forwards screen name to bridge', () => {
        const bridge = new FakeBridge();
        const plugin = makePlugin(bridge);
        plugin.setup(new FakeDelegate());

        plugin.forwardScreen('HomePage');

        expect(bridge.navigatedScreens).toEqual(['HomePage']);
    });
});

// ─── identifier ───────────────────────────────────────────────────────────────

describe('WebEngagePlugin — identifier', () => {
    it('returns "webengage"', () => {
        const plugin = makePlugin(new FakeBridge());
        expect(plugin.identifier).toBe('webengage');
    });
});
